// mdb-angular-ui-kit/esm2022/dropdown/dropdown-menu.directive.mjs
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
// eslint-disable-next-line @angular-eslint/component-class-suffix
export class MdbDropdownMenuDirective {
    elementRef;
    _renderer;
    constructor(elementRef, _renderer) {
        this.elementRef = elementRef;
        this._renderer = _renderer;
    }
    menuPositionClassChanged = new EventEmitter();
    get menuPositionClass() {
        return this._menuPositionClass;
    }
    set menuPositionClass(newClass) {
        const host = this.elementRef.nativeElement;
        const isSameClass = host.classList.contains(newClass);
        if (this._menuPositionClass !== newClass && !isSameClass) {
            const menuPositionClasses = [
                'dropdown-menu-start',
                'dropdown-menu-sm-start',
                'dropdown-menu-md-start',
                'dropdown-menu-lg-start',
                'dropdown-menu-xl-start',
                'dropdown-menu-xxl-start',
                'dropdown-menu-xxl-start',
                'dropdown-menu-xxl-start',
                'dropdown-menu-end',
                'dropdown-menu-sm-end',
                'dropdown-menu-md-end',
                'dropdown-menu-lg-end',
                'dropdown-menu-xl-end',
                'dropdown-menu-xxl-end',
                'dropdown-menu-xxl-end',
                'dropdown-menu-xxl-end',
            ];
            menuPositionClasses.forEach((className) => {
                this._renderer.removeClass(host, className);
            });
            this._renderer.addClass(host, newClass);
            this.menuPositionClassChanged.emit(this.menuPositionClass);
        }
    }
    _menuPositionClass;
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: MdbDropdownMenuDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.4", type: MdbDropdownMenuDirective, selector: "[mdbDropdownMenu]", inputs: { menuPositionClass: "menuPositionClass" }, outputs: { menuPositionClassChanged: "menuPositionClassChanged" }, exportAs: ["mdbDropdownMenu"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: MdbDropdownMenuDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: '[mdbDropdownMenu]',
                    exportAs: 'mdbDropdownMenu',
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i0.Renderer2 }], propDecorators: { menuPositionClassChanged: [{
                type: Output
            }], menuPositionClass: [{
                type: Input
            }] } });
